### PR TITLE
preload: fix path handling when resolution points to kernel fs

### PR DESCRIPTION
**Pull request by marcinslusarz**
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemfile/118)

### CODE REVIEW

**Comment by krzycz** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/libpmemfile/path_resolve.c, line 311 at r1](https://reviewable.io:443/reviews/pmem/pmemfile/118#-KidZ8bCpeyVoEe1gdMb:-KidZ8bCpeyVoEe1gdMc:b-iyfb85) ([raw file](https://github.com/pmem/pmemfile/blob/1af324e8f6758ef50b5b014758af7314b8c3d53d/src/libpmemfile/path_resolve.c#L311)):*
> ```C
> 			result->path[end] = '/';
> 
> 		if (at_root && (end - resolved) == 2 &&
> ```

sometimes, it is a good idea to add some comments...

---

*[src/libpmemfile/path_resolve.c, line 312 at r1](https://reviewable.io:443/reviews/pmem/pmemfile/118#-KidZ0IaZawqaYWvwXyn:-KidZ0IaZawqaYWvwXyo:b3qscrb) ([raw file](https://github.com/pmem/pmemfile/blob/1af324e8f6758ef50b5b014758af7314b8c3d53d/src/libpmemfile/path_resolve.c#L312)):*
> ```C
> 
> 		if (at_root && (end - resolved) == 2 &&
> 				memcmp(&result->path[resolved],	"..", 2) == 0) {
> ```

TAB => space

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/118)*

**Comment by krzycz** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/118#-:-KidZTiPiTyznC6hNaop:bnfp4nl)*

**Comment by marcinslusarz** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/libpmemfile/path_resolve.c, line 311 at r1](https://reviewable.io:443/reviews/pmem/pmemfile/118#-KidZ8bCpeyVoEe1gdMb:-Kido3LO0PejxumLB65B:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/1af324e8f6758ef50b5b014758af7314b8c3d53d/src/libpmemfile/path_resolve.c#L311)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

sometimes, it is a good idea to add some comments...
</blockquote></details>

Done.

---

*[src/libpmemfile/path_resolve.c, line 312 at r1](https://reviewable.io:443/reviews/pmem/pmemfile/118#-KidZ0IaZawqaYWvwXyn:-Kido3lOXUBOJ7HSOYuS:b-896fix) ([raw file](https://github.com/pmem/pmemfile/blob/1af324e8f6758ef50b5b014758af7314b8c3d53d/src/libpmemfile/path_resolve.c#L312)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

TAB => space
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/118)*

**Comment by krzycz** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/118)*

File: src/libpmemfile/path_resolve.c [hunk 1]
```diff
@@ -303,9 +303,25 @@ resolve_path(struct fd_desc at,
...
+		 * syscalls.
+		 */
+		if (at_root && (end - resolved) == 2 &&
+				memcmp(&result->path[resolved], "..", 2) == 0) {
+			exit_pool(result, resolved, &size);
+			at_root = false;
+			continue;
+		}
+
+		at_root = false;
```
Comment (GBuella) [inline src/libpmemfile/path_resolve.c]: Why set it to false here?
What if, e.g.: it was true, and the next component is "."?

File: src/libpmemfile/path_resolve.c [hunk 2]
```diff
@@ -268,6 +266,8 @@ resolve_path(struct fd_desc at,
 	if (path[0] == '/')
 		result->at.pmem_fda.pool = NULL;
 
+	bool at_root = false;
```
Comment (GBuella) [inline src/libpmemfile/path_resolve.c]: Can we call it `at_pmem_root`? Would make things more clear, wouldn't it?

Comment (marcinslusarz) [inline src/libpmemfile/path_resolve.c]: Sure.

File: src/libpmemfile/path_resolve.c (hunk 1 above)
Comment (marcinslusarz) [inline src/libpmemfile/path_resolve.c]: Then the last "else" branch will reset at_root to true.

**Comment by krzycz** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/118)*

**Pull request merged**